storage/dataflux/next_page.go

```diff
@@ -0,0 +1,184 @@
+	"google.golang.org/api/iterator"
+)
+
+const (
+	// wsDefaultPageSize specifies the number of object results to include in a single page for worksteal listing.
+	wsDefaultPageSize = 1000
+)
+
+// nextPageOpts specifies options for next page of listing result .
+type nextPageOpts struct {
+	// startRange is the start offset of the objects to be listed.
+	startRange string
+	// endRange is the end offset of the objects to be listed.
+	endRange string
+	// bucketHandle is the bucket handle of the bucket from which objects are to be listed.
+	bucketHandle *storage.BucketHandle
+	// query is the storage.Query to filter objects for listing.
+	query storage.Query
+	// skipDirectoryObjects is to indicate whether to skip or list directory objects.
+	skipDirectoryObjects bool
+	// generation is the generation number of the last object in the page.
+	generation int64
+}
+
+// nextPageResult represents the results of fetching a single page of objects
+// from a GCS listing operation and information for remaining objects to be listed.
+type nextPageResult struct {
+	// items is the list of objects listed.
+	items []*storage.ObjectAttrs
+	// doneListing indicates whether the lister has completed listing.
+	doneListing bool
+	// nextStartRange is the start offset of the next page of objects to be listed.
+	nextStartRange string
+	// generation is the generation number of the last object in the page.
+	generation int64
+}
+
+// nextPage retrieves a single page of objects from GCS using the provided
+// listing options (nextPageOpts). It returns a nextPageResult containing the
+// list of objects, a flag indicating if the listing is complete, the starting
+// point for the next page, and the generation of the last object in the page.
+// In case multiple versions of objects needs to be listed, then
+func nextPage(ctx context.Context, opts nextPageOpts) (*nextPageResult, error) {
+
+	opts.query.StartOffset = addPrefix(opts.startRange, opts.query.Prefix)
+	opts.query.EndOffset = addPrefix(opts.endRange, opts.query.Prefix)
+	objectIterator := opts.bucketHandle.Objects(ctx, &opts.query)
+	var items []*storage.ObjectAttrs
+
+	// nameLexLast is the name of lexicographically last object in the page.
+	nameLexLast := ""
+	// indexLexLast is the index of lexicographically last object in the page.
+	// If the item is iterated but not added to the items list, then indexLexLast is -1.
+	indexLexLast := 0
+	// indexItemLast is the index of the last item in the items list.
+	indexItemLast := -1
+
+	// The Go Listing API does not expose an interface to list multiple objects together,
+	// thus we need to manually loop to construct a page of results using the iterator.
+	for i := 0; i < wsDefaultPageSize; i++ {
+		attrs, err := objectIterator.Next()
+
+		// If the lister has listed the last item for the assigned range,
+		// then set doneListing to true and return.
+		if err == iterator.Done {
+			return &nextPageResult{
+				items:          items,
+				doneListing:    true,
```

Comment:
I'm confused. It looks like we are discarding results we already got in some cases - if we reach the end of this function without encountering iterator.Done, then we are returning without exhausting objects from `objectIterator`, which may still have items we can get without another call to GCS?

Reply:
So I think this is a case where we are relying on undocumented GCS behaviour - unless you know for sure that GCS fetches 1000 objects every time and that will never change? If not, this code should not assume that GCS will always fetch the same amount of objects.

Reply:
You are right, 100 objects is not documented. It was something that I noticed. 
Its not relying on GCS fetching 1000 objects, but then we might end up discarding some of the objects that are already fetched. I can set maxSize to 1000, that way it will fetch 1000 objects only and we won't end up discarding fetched objects. 
Thanks Brenna.